lib/auth/auth.go

```diff
@@ -6154,6 +6165,171 @@ func (a *Server) CleanupNotifications(ctx context.Context) {
+			identifiersMap[id.Spec.UniqueIdentifier] = true
+		}
+
+		// owners is the combined list of owners for relevant access lists we are creating the notification for.
+		var owners []string
+
+		// Check for access lists which haven't already been accounted for in a notification
+		var needsNotification bool
+		for _, accessList := range relevantLists {
+			if !identifiersMap[accessList.GetName()] {
+				needsNotification = true
+				// Create a unique identifier for this access list so that we know it has been accounted for.
+				if _, err := a.CreateUniqueNotificationIdentifier(ctx, threshold.prefix, accessList.GetName()); err != nil {
+					a.logger.WarnContext(ctx, "failed to create notification identifier", "error", err, "access_list", accessList.GetName())
+					continue
+				}
+				for _, owner := range accessList.Spec.Owners {
+					owners = append(owners, owner.Name)
+				}
+			}
+		}
+		// Deduplicate the owners list.
+		owners = slices.Compact(owners)
+
+		// Create the notification for this reminder treshold for all relevant owners.
+		if needsNotification {
+			_, err := a.Services.CreateGlobalNotification(ctx, &notificationsv1.GlobalNotification{
+				Spec: &notificationsv1.GlobalNotificationSpec{
+					Matcher: &notificationsv1.GlobalNotificationSpec_ByUsers{
+						ByUsers: &notificationsv1.ByUsers{
+							Users: owners,
+						},
+					},
+					Notification: &notificationsv1.Notification{
+						Spec:    &notificationsv1.NotificationSpec{},
+						SubKind: threshold.notificationSubkind,
+						Metadata: &headerv1.Metadata{
+							Labels: map[string]string{types.NotificationTitleLabel: threshold.title},
+						},
+					},
+				},
+			})
+			if err != nil {
+				a.logger.WarnContext(ctx, "Failed to create access list reminder notification", "error", err)
+			}
+
+			// Also create a notification for users who have CRUD permissions for access lists. This is because they can also review access lists.
+			_, err = a.Services.CreateGlobalNotification(ctx, &notificationsv1.GlobalNotification{
+				Spec: &notificationsv1.GlobalNotificationSpec{
+					Matcher: &notificationsv1.GlobalNotificationSpec_ByPermissions{
+						ByPermissions: &notificationsv1.ByPermissions{
+							RoleConditions: []*types.RoleConditions{
+								{
+									Rules: []types.Rule{
+										{
+											Resources: []string{types.KindAccessList},
+											Verbs:     services.RW(),
+										},
+									},
+								},
+							},
+						},
+					},
+					// Exclude the list of owners so that they don't get a duplicate notification, since we already created a notification for them.
+					ExcludeUsers: owners,
+					Notification: &notificationsv1.Notification{
+						Spec:    &notificationsv1.NotificationSpec{},
+						SubKind: threshold.notificationSubkind,
+						Metadata: &headerv1.Metadata{
+							Labels: map[string]string{types.NotificationTitleLabel: threshold.title},
+						},
+					},
+				},
+			})
```

Comment:
This large block distracts from the logic of the function - can we break it out into a helper?